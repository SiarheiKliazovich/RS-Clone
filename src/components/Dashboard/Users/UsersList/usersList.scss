@import "../../../../sass/variables";
@import "../../../../sass/mixins";

.users-list__wrapper {
    display: flex;
    flex-direction: column;
    align-items: center;
    min-width: 800px;
    width: 100%;
}

.users-list__item {
    width: 100%;

    table {
        width: 100%;

        .table-header {
            font-size: $font-size-lg;
            font-weight: $font-weight-bolder;
        }

        th {
            height: 50px;
            vertical-align: middle;
        }

        tr {
            vertical-align: middle;
        }
    }
}

.image-column {
    width: 50px;

    @include media-small-mobile {
        width: 30px;
    }
}

.table-header {
    height: 50px;
    font-size: 22px;
}

.icon {
    background-image: url("../../../../assets/images/png/icon-1.png");
    background-size: contain;
    background-repeat: no-repeat;
    width: 30px;
    height: 35px;
    border: none;
    cursor: pointer;

    @include media-small-mobile {
        width: 20px;
        background-position: center;
    }
}

.button__promo {
    cursor: pointer;
    border-radius: $border-radius-sm;
    border: none;
    padding: 5px 15px;
    color: $color-white1;
    margin-right: 5px;
}
