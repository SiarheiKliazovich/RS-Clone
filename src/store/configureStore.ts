/* eslint-disable @typescript-eslint/ban-ts-comment */
import { applyMiddleware, createStore } from "redux";
import thunkMiddleware from "redux-thunk";

import loggerMiddleware from "./logger";
import rootReducer from "../reducers";

const configureStore = () => {
  const middlewares = [loggerMiddleware, thunkMiddleware];
  const middlewareEnhancer = applyMiddleware(...middlewares);

  const store = createStore(rootReducer, middlewareEnhancer);

  // @ts-ignore:next-line
  if (process.env.NODE_ENV !== "production" && module.hot) {
    // @ts-ignore:next-line
    module.hot.accept("./../reducers", () => store.replaceReducer(rootReducer));
  }

  return store;
};

export default configureStore;
