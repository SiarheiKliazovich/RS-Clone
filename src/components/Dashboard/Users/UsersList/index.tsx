import { FunctionComponent } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useTranslation } from "react-i18next";

import {
  deleteUser,
  getUsers,
  updateUser,
} from "../../../../services/apiUsers";
import {
  userDeleted,
  usersFetching,
  usersFetched,
  usersFetchingError,
} from "../../../../actions";
import { lightTheme, nightTheme } from "../../../../data/constants";
import { IStore } from "../../../../interfaces/store";
import { IAuthReducer } from "../../../../interfaces/authReducer";
import { IUsersListComponent } from "../../../../interfaces/usersListComponent";
import Hover from "../../../Hover";

import "./usersList.scss";

const UsersList: FunctionComponent<IUsersListComponent> = ({
  users,
  activeButton,
}) => {
  const appInterfaceStore = useSelector((state: IStore) => state.appInterface);
  const { isNightMode, isNavbarNightMode, accentColor } = appInterfaceStore;
  const authStore = useSelector((state: IAuthReducer) => state.auth);
  const { user } = authStore;

  const backgroundColor = isNightMode
    ? nightTheme.background.element
    : lightTheme.background.element;

  const dispatch = useDispatch();
  const role = activeButton === "1" ? "USER" : "SELLER";

  const { t } = useTranslation("dataLang");

  const deleteItem = async (id: string) => {
    try {
      await deleteUser(id);
      dispatch(userDeleted(id));
      dispatch(usersFetching());
      const users = await getUsers(role);
      dispatch(usersFetched(users));
    } catch {
      dispatch(usersFetchingError());
    }
  };

  const updateItem = async (id: string) => {
    try {
      await updateUser(id, { role: "MANAGER" });
      dispatch(usersFetching());
      const users = await getUsers(role);
      dispatch(usersFetched(users));
    } catch {
      dispatch(usersFetchingError());
    }
  };

  return (
    <>
      <div className="users-list__wrapper">
        <div className="users-list__item">
          <table>
            <tbody>
              <tr className="table-header">
                <th>#</th>
                <th>{t("usersTable.email")}</th>
                <th>{t("usersTable.id")}</th>
                {user.role === "ADMIN" ? (
                  <>
                    <th className="image-column"></th>
                    <th className="table-promotion"></th>
                  </>
                ) : null}
              </tr>
              {users.map((item, i) => {
                return (
                  <tr key={item._id}>
                    <th>{i + 1}</th>
                    <th>{item.email}</th>
                    <th>{item._id.slice(0, 8).toLocaleUpperCase()}</th>
                    {user.role === "ADMIN" ? (
                      <>
                        <th>
                          <button
                            className="button__delete-order"
                            onClick={() => deleteItem(item._id)}
                          >
                            <i className="fa fa-remove"></i>
                            {t("ordersTable.delete")}
                          </button>
                        </th>
                        <th className="table-promotion">
                          <Hover>
                            <button
                              style={{
                                backgroundColor: isNavbarNightMode
                                  ? backgroundColor
                                  : accentColor.static,
                              }}
                              className="button__promo"
                              onClick={() => updateItem(item._id)}
                            >
                              {t("usersTable.promote")}
                            </button>
                          </Hover>
                        </th>
                      </>
                    ) : null}
                  </tr>
                );
              })}
            </tbody>
          </table>
        </div>
      </div>
    </>
  );
};

export default UsersList;
