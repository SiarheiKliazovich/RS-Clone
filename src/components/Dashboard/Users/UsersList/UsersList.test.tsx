import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import UsersList from ".";

const mockStore = configureStore([]);

describe("UsersList", () => {
  // eslint-disable-next-line
  let store: any;

  const users = [
    {
      _id: "1",
      email: "test1@test.com",
      password: "password",
      role: "role",
      date: "date",
    },
    {
      _id: "2",
      email: "test2@test.com",
      password: "password",
      role: "role",
      date: "date",
    },
    {
      _id: "3",
      email: "test3@test.com",
      password: "password",
      role: "role",
      date: "date",
    },
  ];

  beforeEach(() => {
    store = mockStore({
      appInterface: {
        accentColor: {
          static: "#000",
          hover: "#fff",
        },
      },
      langReducer: {
        lang: "en",
      },
      favouriteTieReducer: {
        addTieLoadingStatus: "load",
      },
      auth: {
        user: {
          role: "USER",
        },
      },
    });
  });

  test("renders correctly", () => {
    const { container } = render(
      <Provider store={store}>
        <UsersList users={users} activeButton="1" />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
