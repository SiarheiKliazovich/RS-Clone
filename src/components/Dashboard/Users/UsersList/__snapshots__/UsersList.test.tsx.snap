// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersList renders correctly 1`] = `
<div>
  <div
    class="users-list__wrapper"
  >
    <div
      class="users-list__item"
    >
      <table>
        <tbody>
          <tr
            class="table-header"
          >
            <th>
              #
            </th>
            <th>
              usersTable.email
            </th>
            <th>
              usersTable.id
            </th>
          </tr>
          <tr>
            <th>
              1
            </th>
            <th>
              test1@test.com
            </th>
            <th>
              1
            </th>
          </tr>
          <tr>
            <th>
              2
            </th>
            <th>
              test2@test.com
            </th>
            <th>
              2
            </th>
          </tr>
          <tr>
            <th>
              3
            </th>
            <th>
              test3@test.com
            </th>
            <th>
              3
            </th>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
